Add proposal about the common libraries concept in Kyma (#143)
kyma-project/community: sig-and-wg/sig-core/proposals/common-libraries.md

Those libraries should not be treated as public and we should not focus on having the generic libraries there. Instead, we should place there all things which are helpful for Kyma developers even if they are tightly coupled with the Kyma domain.

In the future, we should execute the next phase and review our libraries. We should also check if we are ready to extract common libraries outside and announce them as the official SDK for all Kyma internal and external developers. 

### Package name 

In Go ecosystem, there is a pretty popular `pkg` layout pattern. Many well-known projects, such as [Kubernetes](https://github.com/kubernetes/kubernetes/tree/master/pkg), [Docker](https://github.com/moby/moby/tree/master/pkg), [Grafana](https://github.com/grafana/grafana/tree/master/pkg), [Influxdb](https://github.com/influxdata/influxdb/tree/master/pkg), [Etcd](https://github.com/etcd-io/etcd/tree/master/pkg), use this pattern. 
However, the `pkg` directory is used mainly for public libraries. They can be used internally by your application but also by external projects. 
This is an informal contract between you and the external users of your code. 

> In Go, there is an ability to hide code using internal directories. If you put your code in an `internal` directory, no external project will be able to import that code. 

Reviewer: ```suggestion
In Go, there is a possibility to hide code using internal directories. If you put your code in an `internal` directory, no external project will be able to import that code. 
```

> Initially, the idea was to put our private libraries in `kyma/internal/pkg/{private_libs}` but we consolidated all project to one repository, so there is no option to vendor such libraries in our project. In this case, when we execute the `dep ensure` command, we receive such error:  

Reviewer: ```suggestion
Initially, the idea was to put our private libraries in `kyma/internal/pkg/{private_libs}` but we consolidated all the project to one repository, so there is no option to vendor such libraries in our project. In this case, when we execute the `dep ensure` command, we receive such error:  
```

### Next step

The next step is to create a proposal about the best practices. For example, we should get rid of the [`Sleep` functions](https://github.com/kyma-project/kyma/blob/09b945ec0a897822b40666e518f2f258ae66e7e8/tests/acceptance/servicecatalog/wait/wait.go#L8-L24) from our test cases and require a better approach. 
But first, we need a place where we can refer to the proper code which can be used out of the box. In such case, the common libraries can be used.
 
Another example is to unify the logging approach in our projects. This will be also resolved by the other proposal in which the source code will be placed in the common libraries package, so all projects can simply use it.  

## Consequences

Once approved, these are the consequences:
- new package for common libraries must be created and maintained
- new expert CODEOWNERS must be defined for reviewing libraries added for common package
- new package must be announced and other projects should switch to use the official libraries instead of custom implementation

> - developers always need to check our libraries not to reimplement the business logic

Reviewer: ```suggestion
- developers must always check our libraries not to reimplement the business logic
```
